### ISSUE TITLE
JSON Schema

### ISSUE THREAD

**Issue by NathanHazout** [jwadhams/json-logic-js#55]
Does an official JSON Schema (http://json-schema.org/) currently exists for json-logic syntax?

If not, do you think it would be possible?

**Comment by jwadhams** [comment]
Not to my knowledge.
I think you'd end up writing a schema for every supported operation, right? And then any operation can take an operation in almost any argument, but in a few minutes Googling I couldn't find something inheritance-like to say:
{"+":[]} is a type of JsonLogic statement. It takes an array of numbers or "JsonLogic statements" 

Maybe you could write one big schema that takes every supported operation as an optional property, but I'm not sure the format lets you say "you have to have exactly one of this huge property list"... and it doesn't play nice with [adding operations](http://jsonlogic.com/add_operation.html).

What benefit do you expect from having a schema?

**Comment by AndyOGo** [comment]
Yes, indeed an JSON Schema would be awesome 🎉

**Comment by AndyOGo** [comment]
@nasht00 
@jwadhams 
@iddan 

This package is really great, unfortunately it seems that it isn't maintained anymore.
Therefore we forked it, rewrote it to ESnext and [added JSON Schemas](https://github.com/axa-ch/json-logic-js/pull/4).
I just [released `v2.0.0`](https://github.com/axa-ch/json-logic-js/releases/tag/v2.0.0) at on [npm `@axa-ch/json-logic-js`](https://www.npmjs.com/package/@axa-ch/json-logic-js)

**Comment by yelworc** [comment]
Whoa, @AndyOGo – I was just evaluating `json-logic` for a project yesterday, and wondered if there was a schema to validate the resulting logic data – and here it is, added just a day before 😆 

Now if only the schema files were really hosted at the given URLs, my IDE (VS Code) could actually use them for validation 😏

**Comment by AndyOGo** [comment]
@yelworc 
Yeah, you are absolutely right, thats why I want to play the ball back to @jwadhams 
So that all is nicely grouped together and public at jsonlogic.com

**Comment by yelworc** [comment]
Hmm. After playing with it a little more, I'm starting to wonder if this is as helpful as I had hoped. The schema validates _anything_, due to the inclusion of `common/all-types` in `common/any` – which is fair, I guess, since the `apply` function does accept any value as its first parameter, but that makes the whole thing a bit pointless. Am I missing something here?

**Comment by AndyOGo** [comment]
@yelworc 
Yeah, you are absolutely right, actually ideally this would be `oneOf` instead of `anyOf`:
https://github.com/axa-ch/json-logic-js/blob/d56799465c6ce011c23f09d45267bba55ae55ec7/schemas/common/any.json#L8

and maybe it works if `object` is removed from all-types here:
https://github.com/axa-ch/json-logic-js/blob/d56799465c6ce011c23f09d45267bba55ae55ec7/schemas/common/all-types.json#L7

Indeed `jsonLogic.apply` and all it's operands are very forgiving, so I decided to start with a forgiving schema.
But I agree with you it's rather paradox and makes these schemas kinda pointless.
Though you can still replace `all-types` and `any` with your own.

I would also really like to get the inventors feedback @jwadhams 
What do you think?

**Comment by AndyOGo** [comment]
@yelworc 
I just started to refactor the schemas to get rid of `anyOf` and replace it with `oneOf`.
https://github.com/axa-ch/json-logic-js/pull/16

All official tests pass now, so seems to be good.
Only thing is object with single keys which are not a op name is missing 🤔

**Comment by yelworc** [comment]
@AndyOGo cool! This did fix one false negative I saw before (`{ if: [true, 'x', 'y'] }`, for what it's worth).

Agreed, I'd appreciate some feedback by the maintainer, too; e.g. thoughts regarding a potential future major version with less permissive input handling.

(I'm a bit concerned about the general maintenance status of the project, so I'm glad to see the activity in your fork!)

**Comment by AndyOGo** [comment]
@yelworc 
The last thing which is not ideal I see is, that currently all operations can be combined with any other operator. Tracked here: https://github.com/axa-ch/json-logic-js/issues/17

If you see any other issues, please feel free to create a ticket.

**Comment by jwadhams** [comment]
Thank you all for your hard work, this is spectacular. I haven't been able to dedicate as much time to this library as I'd like.

@AndyOGo I'll take a look at your repo and see if we're in a position to merge up. It looks like really great, thoughtful work.

**Comment by AndyOGo** [comment]
@jwadhams 
We just migrated from `draft-04` to `draft-07` with [`v2.2.2`](https://github.com/axa-ch/json-logic-js/blob/master/CHANGELOG.md#222-2020-05-29).

**Comment by Tajnymag** [comment]
A year and a half later, any news?

**Comment by toddbaert** [comment]
@jwadhams Thanks for all your help efforts with this library, and with the Json-Logic project in general.

Is schema-support something you'd still consider? It would be great to have [these schemas](https://github.com/axa-ch/json-logic-js/tree/master/schemas) actually hosted at jsonlogic.com. I imagine this project is a labor of love, so I understand having limited time to devote to it, but I think this is a killer feature.

**Comment by Tiglat-Pileser** [comment]
Has any progress been made on this issue? I have recently been working on a schema for validating json logics components myself but I would prefer to check that against any other attempts out there if there are any.
